### ISSUE TITLE
[UX] Automatically clear caches after manual upgrades

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#5101]
#5099 is the inspiration of this feature request.

In the words of @indigoxela:
> If doing an update via UI, caches do get cleared. At least, that has worked for me several times.

I would like to confirm if that is the case, and if it applies to both core and contrib.

> If doing an update via filesystem (drush, bee, unzip the file, or ftp upload), which means to simply replace the files (core or contrib), caches do not get flushed automatically. Admins have to do that manually. I don't think, there's a possible fix for that. Backdrop does not "observe" the file system for changes - and couldn't.

I think that updating via drush behaves the same as performing the update via the UI, so caches are cleared. I could be wrong about it though, so I would like to confirm that.

In any way, I would like us to automate that, so that people performing manual updates face less WTFs after upgrading. Here's a very simplified version of what I was thinking:

- Backdrop caches the core version, as it's being reported in `bootstrap.inc` (see https://github.com/backdrop/backdrop/blob/1.x/core/includes/bootstrap.inc#L10) - this could be stored in config (`system.core.json` perhaps?) or in a state (`state_set()`).
- A person performs a manual update, by replacing the `/core` directory with a newer version.
- Next time Backdrop loads a page, it detects that the version number in `bootstrap.inc` is now different from the one previously cached, clears caches, and shows a message to inform that caches have been cleared.

**Comment by klonos** [comment]
I would like thoughts on the above recommendation please. Do people have concerns re performance? Do you think that this is a feature worth having in core?

**Comment by indigoxela** [comment]
>> If doing an update via UI, caches do get cleared. At least, that has worked for me several times.

> I would like to confirm if that is the case, and if it applies to both core and contrib.

Yes, that works correctly for both, core and contrib.

> I would like thoughts on the above recommendation please. Do people have concerns re performance?

I'm no expert regarding performance, but there might be concerns. Because this means, the version check has to happen *on every page load*. And if we implement something similar for contrib, potentially a lot of checks have to be done on every page load. Means: reading all info files of all installed contrib modules and comparing to database entries.

But our performance experts for sure can explain that better.

**Comment by klonos** [comment]
> ...the version check has to happen *on every page load*.

Agreed. But the current version is a constant in bootstrap.inc, which IIUC is loaded and available all the time anyway. It is also my understanding that variables loaded in the state storage system are also available at all times (whether the site is in maintenance mode for instance). So we'd basically be comparing 2 numbers that are always available. I could be wrong though 🤷🏼 

> And if we implement something similar for contrib...

Yeah, that could get really hairy really quick, because contrib project versions are retrieved by parsing individual .info files. If we did this only for core though (and assuming that my previous assumptions re core version being available at all times via the bootstrap.inc constant), we'd avoid most situations where caches need to get cleared.

**Comment by quicksketch** [comment]
This idea seems... mostly safe. I agree that it would have a low likelihood of causing a performance impact. The `BACKDROP_VERSION` constant and the values from `state_get()` are available almost every page load. Though when the state system gets initialized depends on that page (the first call to `state_get()`). Common actions like CSS aggregation or loading a form will include a call to `state_get()`.

We could further mitigate the chances of it having any impact by only checking if the user is logged in, and has a certain permission like `administer software updates` (access to update.php). I don't think we'd want to clear caches for anonymous users, since they may be getting the working, cached version of a page.

---

Separate from performance, we should say a thing or two about the _safety_ of automatically clearing the caches on page load.

In my experience, Backdrop upgrades tend to work better after cache clears. Lingering cache values can cause unexpected PHP notices until the first cache clear.

Backdrop regularly includes updates that do not explicitly require running update.php. While Visiting update.php and clicking "Continue" will clear caches, even if no updates are needed, a lot of times administrators will skip visiting update.php because the site status report (and our release notes) don't indicate any updates are needed.

There is some chance that running update.php is _critical_, and clearing the caches may make your site cease to function. However, in the case that we only do this automatic clearing if the user has `administer software updates`, hopefully the administrator would be able to visit update.php right away if a serious problem occurred. We also already try to minimize the occurrence of such situations where the site is unusable until update.php is run.
